### PR TITLE
Various small improvements to Fresnel code

### LINKED ISSUE #191: add test case to test_fresnel that uses a detector in a fresnel system
<a href="https://github.com/mperrin"></img></a> **Issue by [mperrin](https://github.com/mperrin)**
_Friday Sep 30, 2016 at 15:31 GMT_
_Originally opened as https://github.com/mperrin/poppy/issues/191_

----

see #190

### LINKED ISSUE #242: Fresnel propagation detector and setting pixel pitch
<a href="https://github.com/mb2448"></img></a> **Issue by [mb2448](https://github.com/mb2448)**
_Friday Dec 29, 2017 at 21:57 GMT_
_Originally opened as https://github.com/mperrin/poppy/issues/242_

----

Hello,
_python version = 2.7.12, numpy = 1.13.3, poppy = 0.6.1_

This is a two part question from a casual user of _Proper_ and _Poppy_.

1)  I am trying to add a detector to my _FresnelOpticalSystem_, but the following simple code simulating the arago spot breaks:

```
import numpy as np
import poppy
from poppy.poppy_core import PlaneType
import astropy.units as u
import astropy.io.fits as fits

if __name__ == "__main__":
    npix = 1024 # ~512 is minimum to accurately recover the central diffraction spike
    osys = poppy.FresnelOpticalSystem(pupil_diameter = 0.05*u.m, npix = npix, beam_ratio = 0.25)
    osys.add_optic(poppy.CircularAperture(radius=0.025) )
    osys.add_detector(pixelscale=9e-6*u.m, distance = 2*u.m)
```
with **AttributeError: 'FresnelOpticalSystem' object has no attribute 'oversample'**

2) I am confused about how to set up the following (see #191):  I want a circular aperture that fresnel propagates a certain distance, then look at the beam at that distance with a detector of fixed pixel size.  Is the code above the right approach?  I have been avoiding using _FresnelOpticalSystem_ thus far and simply operating on _FresnelWavefront_, but to downsample the beam to the detector pixel scale (9 microns) using POPPY, it seems the only way is to use the _FresnelOpticalSystem_ and add a detector.  Is there another way?

### LINKED ISSUE #251: test function `test_fresnel_FITS_Optical_element` misses the point
<a href="https://github.com/mperrin"></img></a> **Issue by [mperrin](https://github.com/mperrin)**
_Tuesday May 15, 2018 at 23:24 GMT_
_Originally opened as https://github.com/mperrin/poppy/issues/251_

----

Not sure how this got missed before. Despite the name of that function, it appears to be calculating using a plain old OpticalSystem rather than a FresnelOpticalSystem.  "Contents of box do not match label."  ?!?

### PR DESCRIPTION
**Pull request by mperrin**
This PR addresses a handful of misc issues with Fresnel propagation, including:

- Fixes #242, by enabling the use of Detector instances in FresnelOpticalSystems. A new method on FresnelWavefront will resample a given wavefront onto the desired pixel scale and number of pixels to match the specified detector. To help enable this, Detector pixel scales can now optionally be specified in linear units of e.g. microns/pixel, as well as in angular units like arcsec/pixel. You have to use the linear units for Fresnel detectors. 
- Fixes #191, the lack of a unit test for Detector instances in FresnelOpticalSystems
- Fixes #251, the problem that the test case for "FITSOpticalElement in FresnelOpticalSystem" did not actually use a FresnelOpticalSystem. During the course of fixing this I discovered an edge-case bug in the FITSOpticalElement get_phasor code (didn't properly handle the case of a matching pixel scale but different array shapes between optic and wavefront), so I fixed that too.
- Some small additions to the Fresnel docs page. 
- A handful of misc code cleanup (removal of some now-unneeded Python 2 stuff  in the unit test code (see #262), some duplicated imports, etc.)

### CODE REVIEW

**Comment by mperrin** [comment]
Attention @douglase, this PR will probably be of interest to you. I would tag you as a reviewer, but Github doesn't seem to let me do that for usernames outside of the spacetelescope organization in this case. It would certainly be appreciated if you have time to look this over, but I don't think it needs a detailed review. It's all relatively straightforward fixes to small issues.

File: poppy/fresnel.py [hunk 1]
```diff
@@ -958,6 +959,52 @@ def apply_lens_power(self, optic, ignore_wavefront=False):
...
+        _log.debug("Desired detector pixel scale: {}".format(detector.pixelscale))
+        _log.debug("Desired detector FOV: {} pixels, {}".format(detector.shape,
+                                                                      detector.shape[0]*u.pixel*detector.pixelscale,
+                                                                      ))
+
+        # TODO the following works but is not optimal for performance
+        # We should consider cropping out an appropriate subregion prior to performing the zoom.
+        # That makes a difference if the detector is only sampling a small part of a much larger wavefront
+
+        new_wf_real = scipy.ndimage.zoom(self.wavefront.real, pixscale_ratio)
```
Comment (douglase) [inline poppy/fresnel.py]:
```suggestion
        new_wf_real = scipy.ndimage.zoom(self.wavefront.real, pixscale_ratio, order=self.interp_order)
```
Configuration of the interpolation order is useful for special cases, such where you want to keep masks binary

Comment (douglase) [reply poppy/fresnel.py]: see https://github.com/spacetelescope/poppy/blob/e97c97f01a002a71c86b07b1ce2eb8d473575f84/poppy/poppy_core.py#L1924

File: poppy/fresnel.py [hunk 2]
```diff
@@ -958,6 +959,52 @@ def apply_lens_power(self, optic, ignore_wavefront=False):
...
+        new_wf_imag = scipy.ndimage.zoom(self.wavefront.imag, pixscale_ratio)
```
Comment (douglase) [inline poppy/fresnel.py]: same as above

File: poppy/fresnel.py (hunk 1 above)
Comment (mperrin) [reply poppy/fresnel.py]: Turns out this needs to be `order=detector.interp_order` rather than `self`. But yes in any case good suggestion.

**Comment by mperrin** [comment]
ok that's weird, the Travis build has failed to complete. It didn't actually fail the tests, it just errorred out and didn't operate properly. May be a test infrastructure problem. I'm going to restart the test and see if that kicks it into working.

**Comment by mperrin** [comment]
I'm going to go ahead and merge this, regardless of the Travis problem. I don't see how any of these changes can be related to that. The Travis build is not failing, it's just not even operating correctly.
